lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/core/expression/ExpressionScope.kt

```diff
@@ -513,116 +480,76 @@ public interface ExpressionScope {
+  > internal constructor(internal val label: Expression, internal val output: Output)
 
   /** Create a [MatchBranch], see [match] */
-  public infix fun <Output> String.then(output: Expression<Output>): MatchBranch<String, Output> =
-    MatchBranch(const(this), output)
+  public infix fun <Output : Expression> String.then(
+    output: Output
+  ): MatchBranch<StringExpression, Output> = MatchBranch(const(this), output)
 
   /** Create a [MatchBranch], see [match] */
-  public infix fun <Output> Number.then(output: Expression<Output>): MatchBranch<Number, Output> =
-    MatchBranch(const(this.toFloat()), output)
+  public infix fun <Output : Expression> Number.then(
+    output: Output
+  ): MatchBranch<NumberExpression, Output> = MatchBranch(const(this.toFloat()), output)
 
   /** Create a [MatchBranch], see [match] */
   @JvmName("stringsThen")
-  public infix fun <Output> List<String>.then(
-    output: Expression<Output>
-  ): MatchBranch<String, Output> = MatchBranch(Expression.ofList(this.map { const(it) }), output)
+  public infix fun <Output : Expression> List<String>.then(
+    output: Output
+  ): MatchBranch<StringExpression, Output> =
+    MatchBranch(ExpressionImpl.ofList(this.map(::const)), output)
 
   /** Create a [MatchBranch], see [match] */
   @JvmName("numbersThen")
-  public infix fun <Output> List<Number>.then(
-    output: Expression<Output>
-  ): MatchBranch<Number, Output> =
-    MatchBranch(Expression.ofList(this.map { const(it.toFloat()) }), output)
+  public infix fun <Output : Expression> List<Number>.then(
+    output: Output
+  ): MatchBranch<NumberExpression, Output> =
+    MatchBranch(ExpressionImpl.ofList(this.map { const(it.toFloat()) }), output)
 
   /**
    * Evaluates each expression in [values] in turn until the first non-null value is obtained, and
    * returns that value.
    */
-  public fun <T> coalesce(vararg values: Expression<T?>): Expression<T> =
-    callFn("coalesce", *values)
-
-  /** Returns whether this expression is equal to [other]. */
-  @JvmName("eqBoolean")
-  public infix fun Expression<Boolean>.eq(other: Expression<Boolean>): Expression<Boolean> =
-    callFn("==", this, other)
-
-  /** Returns whether this expression is equal to [other]. */
-  @JvmName("eqString")
-  public infix fun Expression<String>.eq(other: Expression<String>): Expression<Boolean> =
-    callFn("==", this, other)
-
-  /** Returns whether this expression is equal to [other]. */
-  @JvmName("eqNumber")
-  public infix fun Expression<Number>.eq(other: Expression<Number>): Expression<Boolean> =
-    callFn("==", this, other)
+  public fun <T : Expression> coalesce(vararg values: T): T = callFn("coalesce", *values)
 
   /** Returns whether this expression is equal to [other]. */
-  @JvmName("eqDp")
-  public infix fun Expression<Dp>.eq(other: Expression<Dp>): Expression<Boolean> =
-    callFn("==", this, other)
+  public infix fun Expression.eq(other: Expression): BooleanExpression = callFn("==", this, other)
```

Comment:
The actual limitation for comparisons is that the left hand side and right hand side **must** be of the same type according to the style definition ([source](https://github.com/maplibre/maplibre-style-spec/blob/main/src/expression/definitions/comparison.ts#L96C1-L96C9)). That's why I had a different overload for the different types, to ensure that.

That holds true for the other comparisons, too, by the way. `4 > "hello"` is considered invalid, too.

Also, why only Boolean, String, Number? Because equality is defined as strict equality (see [source](https://github.com/maplibre/maplibre-style-spec/blob/main/src/expression/definitions/comparison.ts#L29)), so two lists, or maps, or colors can never be equal, even if they have the same content.